kep/42-podgroup-coscheduling/README.md

```diff
@@ -0,0 +1,125 @@
+	// if there's not enough resources to start all tasks, the scheduler
+	// will not start anyone.
+	MinMember uint32 `json:"minMember"`
+
+	// MaxScheduleTime defines the maximal time of members/tasks to wait before run the pod group;
+	MaxScheduleTime *metav1.Duration `json:"maxScheduleTime,omitempty"`
+}
+
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+	// Current phase of PodGroup.
+	Phase PodGroupPhase `json:"phase"`
+
+	// OccupiedBy marks the podgroup occupied by which group.
+	// Owner reference would be used to filled it, if not initialize, it is empty
+	OccupiedBy string `json:"occupiedBy,omitempty"`
+
+	// The number of actively running pods.
+	// +optional
+	Scheduled uint32 `json:"scheduled"`
+
+	// The number of actively running pods.
+	// +optional
+	Running uint32 `json:"running"`
+
+	// The number of pods which reached phase Succeeded.
+	// +optional
+	Succeeded uint32 `json:"succeeded"`
+
+	// The number of pods which reached phase Failed.
+	// +optional
+	Failed uint32 `json:"failed"`
+
+	// ScheduleStartTime of the group
+	ScheduleStartTime metav1.Time `json:"scheduleStartTime"`
+}
+```
+
+### Controller
+
+We define a controller to reconcile PodGroup status, and we can query the job status through describing the PodGroup. Onece a pod in a group failed, the Group Status is marked Failed. Controller would also help recover from abnormal cases, e.g. batch scheduling is interrupted due to
+cluster upgrade.
+
+### Extension points
+
+#### QueueSort
+
+To make sure a group of pods can be scheduled as soon as possible. We implemented this `extension point`. The main progress is as follows:
+
+1. Sort based on Pod Priority
+2. Pod does not belong to any group
+3. PodGroup creation time
+4. Pod creation time
+
+#### PreFilter
+
+This extension helps pre-filter pods. It is useful, especially when there are not enough resources in a cluster. The overall flow works as below:
+
+1. Allow the pods that do not belong to any group.
+2. If there are no groups scheduling, we check resource, if enough, we allow the pod.
+3. If there are groups running, we check if the current pod belong the group having the max progress(num(Pods)/minMember), if it is, we allow it.
+4. Otherwise, we check if the max finished group can still run when allow this pod. If we can, allow it.
+5. Otherwise, we check if the pod has higher priority compared with the max finished one. If yes, we reject the pod belongs to the group and allow the current one. 
+
+Any pod rejected to run, their group would be added to a denied list with a ttl.
+
+#### Permit
+
+1. When number of pods cannot meet the `minMember` defines in the PodGroup, `Wait` is returned. They will be added to cache with TLL(equal to MaxScheduleTime).
+2. When number meet that, we would send a signal to permit the pods waiting. 
+
+We can define `MaxScheduleTime` for a PodGroup. If anyone of the pods times out, the whole group would be rejected.
```

Comment:
FYI: it's now defined in the latest codebase, and will be available in `master` branch until #45 gets reviewed.

Reply:
I think we should keep this.  Our original version supports from both args and the crd.  If the crd did not set the time, the args defines will work.

scene is as bellow:
A job need 10 pods may time out time after 5s, but a job need 10000 pods would like it to be 5min.

Reply:
I think it is better not to let the user to set the `MaxScheduleTime`. Users will prefer to set the `MaxScheduleTime` Greate than it needed to reserve the resource. This will lead the resource to be reserved for a long time.

> A job need 10 pods may time out time after 5s, but a job need 10000 pods would like it to be 5min.
---
We talked about it before. We can use a factor to compute the `WaitingTime`. The factor can be related to the number of pod.  I add Todo in code before.

Reply:
But for different cluster,  the factor would also be different. 
e.g. 100 nodes or 2000nodes

Reply:
Yes, the size of cluster is also a factor to be considered with the the number of pod. This can be discussed.
And I am still worried about it. @Huang-Wei @cwdsuzhou 
> I think it is better not to let the user to set the MaxScheduleTime. Users will prefer to set the MaxScheduleTime greater than it needed to reserve the resource. This will lead the resource to be reserved for a long time.